Update agent formula to choose an m1/intel tarball based on the host computer

yob opened the pull request:
Homebrew is working through the complexity of transitioning from Intel to Apple Silicon and much is up in the air.

However, our agent formula is quite simple. On Apple Silicon machines, we can ignore whether homebrew itself is running natively and just always download the Apple Silicon version of the agent.

On non Apple Silicon machines, we can ignore whether homebrew itself is running natively and just always download the Intel version of the agent.

This requires some helpers added to homebrew in https://github.com/Homebrew/brew/pull/7995. Will they be available everywhere?

Before merging this we also need to consider the agent CD scripts that update this file automatically: https://github.com/buildkite/agent/blob/b2cf4e3d90f5b155505ec6eb53e007a4d88239d7/.buildkite/steps/release-homebrew.sh

ticky commented:
I think we should follow what Homebrew is planning to do, which is installing based on the flavour of Homebrew installation, not the host CPU specifically.

It is expected that a user could have two Homebrew installations, one amd64 and one arm64, and that each must only install binaries of its flavour. I'm hoping to chase up the shape of this tooling as it becomes supported.

ticky commented:
Looks like what we might want is [`Hardware::CPU.arm?`](https://github.com/Homebrew/homebrew-core/pull/65670/files#diff-df00e346e6e2293ca1363f7b202030ab4e24569f520a49635bc29692e7eb5283R87) instead.

ticky commented on buildkite-agent.rb:
```diff
@@ -4,8 +4,13 @@ class BuildkiteAgent < Formula
     version "3.26.0"
-    url     "https://github.com/buildkite/agent/releases/download/v3.26.0/buildkite-agent-darwin-amd64-3.26.0.tar.gz"
-    sha256  "8a2f78aad0791ebe41a546a6c17b71560852a439ba311d541d63073410645b2d"
+    if Hardware::CPU.physical_cpu_arm64?
```
```suggestion
    if Hardware::CPU.arm?
```

yob commented:
I've been using macos and homebrew for less than 24 hours and you have somewhat closer connections than me, so we should 100% go with your suggestion :joy:

ticky commented:
Following this up with a hopefully-matching PR for the agent-release gear buildkite/agent#1346. Need to work out how to give that a run-through

yob commented:
I've tested this on a m1 mac and it works great.

First. I have `brew` installed to `/opt/homebrew`.

```
jh@Jamess-Air homebrew % cd $(brew --repo buildkite/buildkite)
jh@Jamess-Air homebrew-buildkite % git checkout  detect-arch
Branch 'detect-arch' set up to track remote branch 'detect-arch' from 'origin'.
Switched to a new branch 'detect-arch'

jh@Jamess-Air homebrew-buildkite % brew reinstall buildkite-agent
Warning: You are running macOS on a arm64 CPU architecture.
We do not provide support for this (yet).
Reinstall Homebrew under Rosetta 2 until we support it.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
unsupported configuration.

==> Downloading https://github.com/buildkite/agent/releases/download/v3.26.0/buildkite-agent-darwin-arm64-3.26.0.tar.gz
Already downloaded: /Users/jh/Library/Caches/Homebrew/downloads/2f0c5e21e2383b2c0b3d2c58a6ca143314937b15ff48d403d252270832210e11--buildkite-agent-darwin-arm64-3.26.0.tar.gz
==> Reinstalling buildkite/buildkite/buildkite-agent 
Ignoring existing config file at /opt/homebrew/etc/buildkite-agent/buildkite-agent.cfg
For changes see the updated dist copy at /opt/homebrew/Cellar/buildkite-agent/3.26.0/share/buildkite-agent/buildkite-agent.dist.cfg
==> Caveats
buildkite-agent is now installed!

Don't forget to update your configuration file with your agent token

Configuration file (to configure agent token, meta-data, priority, name, etc):
    /opt/homebrew/etc/buildkite-agent/buildkite-agent.cfg

Hooks directory (for customising the agent):
    /opt/homebrew/etc/buildkite-agent/hooks

Builds directory:
    /opt/homebrew/var/buildkite-agent/builds

Log paths:
    /opt/homebrew/var/log/buildkite-agent.log
    /opt/homebrew/var/log/buildkite-agent.error.log

If you set up the LaunchAgent, set your machine to auto-login as
your current user. It's also recommended to install Caffeine
(http://lightheadsw.com/caffeine/) to prevent your machine from going to
sleep or logging out.

To run multiple agents simply run the buildkite-agent start command
multiple times, or duplicate the LaunchAgent plist to create another
that starts on login.

To have launchd start buildkite/buildkite/buildkite-agent now and restart at login:
  brew services start buildkite/buildkite/buildkite-agent
Or, if you don't want/need a background service you can just run:
  buildkite-agent start
==> Summary
🍺  /opt/homebrew/Cellar/buildkite-agent/3.26.0: 5 files, 19.0MB, built in 1 second
jh@Jamess-Air homebrew-buildkite % lipo -archs $(which buildkite-agent)
arm64
```

I then installed a parallel version of homebrew to `/usr/local` ([using these instructions](https://soffes.blog/homebrew-on-apple-silicon)):

```
% cat ~/.zshrc
export PATH=/opt/homebrew/bin:"${PATH}"
alias ibrew='arch -x86_64 /usr/local/bin/brew'
```

I then install the agent for ibrew:

```
jh@Jamess-Air homebrew-buildkite % pwd
/usr/local/Homebrew/Library/Taps/buildkite/homebrew-buildkite

% git checkout detect-arch
Branch 'detect-arch' set up to track remote branch 'detect-arch' from 'origin'.
Switched to a new branch 'detect-arch'

jh@Jamess-Air homebrew-buildkite % ibrew reinstall buildkite-agent
==> Downloading https://github.com/buildkite/agent/releases/download/v3.26.0/buildkite-agent-darwin-amd64-3.26.0.tar.gz
Already downloaded: /Users/jh/Library/Caches/Homebrew/downloads/d8f9cc08a5127cfccfff527ed77a8992ee5c1b51860f6d13f7d1a887b7e6ba1b--buildkite-agent-darwin-amd64-3.26.0.tar.gz
==> Reinstalling buildkite/buildkite/buildkite-agent 
Ignoring existing config file at /usr/local/etc/buildkite-agent/buildkite-agent.cfg
For changes see the updated dist copy at /usr/local/Cellar/buildkite-agent/3.26.0/share/buildkite-agent/buildkite-agent.dist.cfg
==> Caveats
buildkite-agent is now installed!

Don't forget to update your configuration file with your agent token

Configuration file (to configure agent token, meta-data, priority, name, etc):
    /usr/local/etc/buildkite-agent/buildkite-agent.cfg

Hooks directory (for customising the agent):
    /usr/local/etc/buildkite-agent/hooks

Builds directory:
    /usr/local/var/buildkite-agent/builds

Log paths:
    /usr/local/var/log/buildkite-agent.log
    /usr/local/var/log/buildkite-agent.error.log

If you set up the LaunchAgent, set your machine to auto-login as
your current user. It's also recommended to install Caffeine
(http://lightheadsw.com/caffeine/) to prevent your machine from going to
sleep or logging out.

To run multiple agents simply run the buildkite-agent start command
multiple times, or duplicate the LaunchAgent plist to create another
that starts on login.

To have launchd start buildkite/buildkite/buildkite-agent now and restart at login:
  brew services start buildkite/buildkite/buildkite-agent
Or, if you don't want/need a background service you can just run:
  buildkite-agent start
==> Summary
🍺  /usr/local/Cellar/buildkite-agent/3.26.0: 5 files, 34.7MB, built in 2 seconds

jh@Jamess-Air homebrew-buildkite % lipo -archs /usr/local/bin/buildkite-agent                      
x86_64
```

success!

yob commented:
@ticky do we need to consider whether `Hardware::CPU.arm?` is available on all users homebrew installs? Or will homebrew update itself and ensure that helpers like that are always available when the formula runs?

ticky commented:
> @ticky do we need to consider whether `Hardware::CPU.arm?` is available on all users homebrew installs? Or will homebrew update itself and ensure that helpers like that are always available when the formula runs?

Homebrew generally expects to be auto-updating both the code and the taps in lockstep, it’s possible to turn it off, but it turns both types of update off, so I think assuming the API will be there is entirely reasonable.

yob commented:
sweet :+1:
